packages/opentelemetry-exporter-collector/src/transform.ts

```diff
@@ -0,0 +1,241 @@
+  if (keys.length > maxAttributes) {
+    droppedAttributesCount = keys.length - maxAttributes;
+  }
+
+  for (let i = 0; i <= countKeys - 1; i++) {
+    const key = keys[i];
+    const eventAttributeValue = toCollectorEventValue(
+      attributes && attributes[key]
+    );
+    if (eventAttributeValue) {
+      attributeMap[key] = eventAttributeValue;
+    }
+  }
+
+  return {
+    droppedAttributesCount,
+    attributeMap,
+  };
+}
+
+/**
+ * convert event value
+ * @param value event value
+ */
+export function toCollectorEventValue(
+  value: unknown
+): collectorTypes.AttributeValue | undefined {
+  const attributeValue: collectorTypes.AttributeValue = {};
+
+  if (typeof value === 'string') {
+    attributeValue.stringValue = toCollectorTruncatableString(value);
+  } else if (typeof value === 'boolean') {
+    attributeValue.boolValue = value;
+  } else if (typeof value === 'number') {
+    if (Math.floor(value) === value) {
+      attributeValue.intValue = value;
+    } else {
+      attributeValue.doubleValue = value;
+    }
+  }
+
+  return attributeValue;
+}
+
+/**
+ * convert events
+ * @param events array of events
+ * @param maxAttributes - maximum number of event attributes to be converted
+ */
+export function toCollectorEvents(
+  events: TimedEvent[],
+  maxAttributes: number = OT_MAX_ATTRIBUTES
+): collectorTypes.TimeEvents {
+  let droppedAnnotationsCount = 0;
+  let droppedMessageEventsCount = 0; // not counting yet as messageEvent is not implemented
+
+  const timeEvent: collectorTypes.TimeEvent[] = events.map(
+    (event: TimedEvent) => {
+      let attributes: collectorTypes.Attributes | undefined;
+
+      if (event && event.attributes) {
+        attributes = toCollectorAttributes(event.attributes, maxAttributes);
+        droppedAnnotationsCount += attributes.droppedAttributesCount || 0;
+      }
+
+      let annotation: collectorTypes.Annotation = {};
+      if (event.name || attributes) {
+        annotation = {};
+      }
+
+      if (event.name) {
+        annotation.description = toCollectorTruncatableString(event.name);
+      }
+
+      if (typeof attributes !== 'undefined') {
+        annotation.attributes = attributes;
+      }
+
+      // const messageEvent: MessageEvent;
+
+      const timeEvent: collectorTypes.TimeEvent = {
+        time: hrTimeToTimeStamp(event.time),
+        // messageEvent,
+      };
+
+      if (annotation) {
+        timeEvent.annotation = annotation;
+      }
+
+      return timeEvent;
+    }
+  );
+
+  return {
+    timeEvent,
+    droppedAnnotationsCount,
+    droppedMessageEventsCount,
+  };
+}
+
+/**
+ * determines the type of link, only parent link type can be determined now
+ * @param span
+ * @param link
+ */
+export function toCollectorLinkType(
+  span: ReadableSpan,
+  link: Link
+): collectorTypes.LinkType {
+  const linkSpanId = link.spanContext.spanId;
+  const spanParentId = span.parentSpanId;
+
+  if (linkSpanId === spanParentId) {
```

Comment:
Could you help me understand this logic a bit? My understanding is that trace links refer to trace IDs and span IDs outside of the trace itself

Reply:
```/** The linked span is a parent of the current span. */
export type LinkTypeParentLinkedSpan = 2;```
so if certain span has any link and id of this link will be the same as parentId of that span, then we have `LinkTypeParentLinkedSpan`

Reply:
My understanding is that links should be explicitly specified because they represent additional relationships that go beyond the normal parent-child relationships. See this [Open Census doc about linked traces](https://opencensus.io/tracing/span/link/) that says "A link describes a *cross-relationship* between spans in either the same or different trace" (emphasis mine).

Since the links are about cross-relationships within a trace or about parent/child relationships outside of a trace, I personally don't think it makes sense to try to auto-detect it. We should only save what was explicitly specified.

Or am I missing something here in how I'm thinking about it?

cc/ @mayurkale22

Reply:
so far there is no such information 
this is how the link is defined now in sdk, there is no xtrainformation, only `SpanContext`
https://github.com/open-telemetry/opentelemetry-js/blob/039206493564b4e5858dc883467686385e04317f/packages/opentelemetry-types/src/trace/link.ts#L25

Reply:
https://github.com/open-telemetry/opentelemetry-js/issues/568

Reply:
OK, then maybe for now just leave the link type field blank and then add it later once the SDK supports adding more metadata about links?

Reply:
The allowed types are those 3 only, so I'm returning now "LINK_TYPE_UNSPECIFIED" which is 0. I just added `LINK_TYPE_PARENT_LINKED_SPAN` in case this is detectable based on parentId only.

Reply:
Well, could you make the detection also check that the trace ID matches then? I think that would be a requirement too since links can go outside of a trace.

Could you also add a TODO item that we should change this to use the explicitly specified link type once that's available?